plugins/SegmentEditor/SegmentQueryDecorator.php

```diff
@@ -0,0 +1,62 @@
+<?php
+/**
+ * Piwik - free/libre analytics platform
+ *
+ * @link http://piwik.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ */
+
+namespace Piwik\Plugins\SegmentEditor;
+
+use Piwik\Access;
+use Piwik\DataAccess\LogQueryBuilder;
+use Piwik\Segment\SegmentExpression;
+
+/**
+ * Decorates segment sub-queries in archiving queries w/ the idSegment of the segment, if
+ * a stored segment exists.
+ *
+ * This class is configured for use in SegmentEditor's DI config.
+ */
+class SegmentQueryDecorator extends LogQueryBuilder
+{
+    /**
+     * @var API
+     */
+    private $segmentEditorApi;
+
+    public function __construct(API $segmentEditorApi)
+    {
+        $this->segmentEditorApi = $segmentEditorApi;
+    }
+
+    public function getSelectQueryString(SegmentExpression $segmentExpression, $select, $from, $where, $bind, $groupBy,
+                                         $orderBy, $limit)
+    {
+        $result = parent::getSelectQueryString($segmentExpression, $select, $from, $where, $bind, $groupBy, $orderBy,
+            $limit);
+
+        $idSegments = $this->getSegmentIdOfExpression($segmentExpression);
+        if (!empty($idSegments)) {
+            $result['sql'] = "/* idSegments = [" . implode(', ', $idSegments) . "] */\n" . $result['sql'];
+        }
+
+        return $result;
+    }
+
+    private function getSegmentIdOfExpression(SegmentExpression $segmentExpression)
+    {
+        $segmentEditorApi = $this->segmentEditorApi;
+        $allSegments = Access::doAsSuperUser(function () use ($segmentEditorApi) {
```

Comment:
Yes, I am going through the API for the cache. I agree it's weird to use `Access::doAsSuperUser` here, but I couldn't think of another way to do it w/o moving the caching to an improper place.

The Model class shouldn't contain the caching, since it's just a DAO and the appropriate place to me, is in the API, since that is essentially the HTTP facing portion of Piwik. Perhaps there should be another layer in between API & DAO (so the API would always defer to it and only do access-control, input mapping).

Actually, I think that would be better. But such a concept should be a normal Piwik development concept. I will talk to you on slack about this.

Anyway, if you have a specific suggestion let me know, but I agree using doAsSuperUser should be avoided.